akka-stream/src/main/scala/akka/stream/io/Timeouts.scala

```diff
@@ -0,0 +1,222 @@
+package akka.stream.io
+
+import java.util.concurrent.TimeoutException
+
+import akka.actor.{ Cancellable, ActorSystem }
+import akka.stream.{ FlowShape, Outlet, Inlet, BidiShape }
+import akka.stream.scaladsl.{ BidiFlow, Flow }
+import akka.stream.stage.{ OutHandler, InHandler, GraphStageLogic, GraphStage }
+
+import scala.concurrent.duration.{ Deadline, FiniteDuration }
+
+/**
+ * Various stages for controlling timeouts on IO related streams (although not necessarily).
+ *
+ * The common theme among the processing stages here that
+ *  - they wait for certain event or events to happen
+ *  - they have a timer that may fire before these events
+ *  - if the timer fires before the event happens, these stages all fail the stream
+ *  - otherwise, these streams do not interfere with the element flow, ordinary completion or failure
+ */
+object Timeouts {
+
+  /**
+   * If the first element has not passed through this stage before the provided timeout, the stream is failed.
+   */
+  // FIXME: Where should system come from? (Actually we need a Scheduler and ExecutionContext)
+  def initalTimeout[T](timeout: FiniteDuration, system: ActorSystem): Flow[T, T, Unit] =
+    Flow.wrap(new InitialTimeout[T](system, timeout))
+
+  /**
+   * If the completion of the stream does not happen until the provided timeout, the stream is failed.
+   */
+  // FIXME: Where should system come from? (Actually we need a Scheduler and ExecutionContext)
+  def completionTimeout[T](timeout: FiniteDuration, system: ActorSystem): Flow[T, T, Unit] =
+    Flow.wrap(new CompletionTimeout[T](system, timeout))
+
+  /**
+   * If the time between two processed elements exceed the provided timeout, the stream is failed.
+   */
+  // FIXME: Where should system come from? (Actually we need a Scheduler and ExecutionContext)
+  def idleTimeout[T](timeout: FiniteDuration, system: ActorSystem): Flow[T, T, Unit] =
+    Flow.wrap(new IdleTimeout[T](system, timeout))
+
+  /**
+   * If the time between two processed elements *in any direction* exceed the provided timeout, the stream is failed.
+   *
+   * There is a difference between this stage and having two idleTimeout Flows assembled into a BidiStage.
+   * If the timout is configured to be 1 seconds, then this stage will not fail even though there are elements flowing
+   * every second in one direction, but no elements are flowing in the other direction. I.e. this stage considers
+   * the *joint* frequencies of the elements in both directions.
+   */
+  // FIXME: Where should system come from? (Actually we need a Scheduler and ExecutionContext)
```

Comment:
I'm all for exposing just `schedule`/`scheduleOnce` instead of anything akka specific, in fact that's how I wanted to impl the AsyncStage scheduling: https://github.com/akka/akka/pull/18225/files#diff-9b5ba5308f1cef1574d15d27aa125b9bR663 :)